src/TableFunctions/TableFunctionS3.cpp

```diff
@@ -176,12 +177,12 @@ void TableFunctionS3::parseArgumentsImpl(ASTs & args, const ContextPtr & context
             configuration.auth_settings.secret_access_key = checkAndGetLiteralArgument<String>(args[args_to_idx["secret_access_key"]], "secret_access_key");
 
         configuration.auth_settings.no_sign_request = no_sign_request;
+
+        if (configuration.format == "auto")
+            configuration.format = FormatFactory::instance().getFormatFromFileName(Poco::URI(url).getPath(), true);
     }
 
     configuration.keys = {configuration.url.key};
-
-    if (configuration.format == "auto")
-        configuration.format = FormatFactory::instance().getFormatFromFileName(Poco::URI(configuration.url.uri.getPath()).getPath(), true);
```

Comment:
The problem was here in `Poco::URI(configuration.url.uri.getPath()).getPath()`. Initially it was a workaround to extract correct filename from url with query parameters, because S3::URI encoded `?` symbol. Now I just use initial url here, so nothing is encoded and `getPath()` method returns path without query parameters